### DISCUSSION TITLE
UX options for a real-smooth, non-clunky custom-language REPL

### DISCUSSION THREAD

**Discussion by metaleap** [microsoft/vscode-discussions#1232] [Extension Development QnA] [answered]
I'm pondering which of the built-in GUI affordances is the best choice for offering a REPL in one's custom language extension. The REPL result responses (other than errors) will be in that language's own syntax.

- **Chat Extensions / Language Models:** despite the REPL backend expecting proper syntax and not natural language, would probably be real neat and good UX (REPLs are chats and vice versa, in primitive essence), but seemingly it's all restricted to Copilot / GPTs — even if it weren't so, might not have syntax-coloring for inputs and outputs, or auto-complete
- **Web View:** will likely easily feel non-native / clunky, hassle to CSS it (and keep updated with the latest VSC stylesheet), extra hassle to syntax-color the REPL inputs and responses, even more hassle to offer auto-complete
- **InputBox-MessageBox roundtripping:** nope :grin:
- **Debug Adapter Protocol:** also poor REPL UX with the Debug Console (syntax coloring, autocomplete etc) — also potentially confusing/diluting since there's no real debug experience (breakpoints, stepping, etc)
- **Notebooks:** seems like the ideal candidate to use, but apparently that API assumes-requires an underlying "notebook file" in the workspace. Would be nice if it weren't so, but I guess users will just have to bear that (unless I can fix it to a custom FileSystemProvider that just no-ops), even though the file will not contain anything at all (what with a live REPL session, communicated over custom LSP requests/responses, starting from a "blank slate" by design every time the LSP starts).
- **Custom document type:** might also be an idea vaguely, but I guess the "input" is always the document, end user needing to manually clear it for their next request input (and the response output goes — who-knows-where...)

Any opinions, or other ideas, or options I have overlooked / not found?

**Comment by isidorn** [comment, marked as answer]
This is currently not supported. But is a fair feature request that we can keep open here to get more community feedback. Thanks

**Reply by metaleap** [reply]
I'll see how it flows with the Notebooks capabilities with the wip [Gerbil LSP](https://github.com/metaleap/gerbil-lsp/) and update here in a couple weeks. Given they have multi-language syntax coloring & intellisense in the code cells, could be pretty superb actually. We'll see about the fluidity..

**Reply by isidorn** [reply]
That approach sounds reasonable to me.

**Comment by metaleap** [comment]
Update on Notebooks UX: for now [I'm seeing **none** of the otherwise-fully-working custom-language stuff](https://github.com/microsoft/vscode-discussions/discussions/1251) actually surface in (properly-languaged) Notebook input cells (except for syntax colors ie tmtheme grammar, that works), to my dismayed surprise — I have fully-working (dummy) hovers, (dummy) completions, etc. operating fine in normal VSCode text editors, how come using the same lang in Notebook input cells brings in none of that? Must be missing sth..
